Sources/performance-worker/BenchmarkService.swift

```diff
@@ -0,0 +1,185 @@
+        $0.payload = Grpc_Testing_Payload.with {
+          $0.body = Data(count: responseSize)
+        }
+      }
+    )
+  }
+
+  /// Single-sided unbounded streaming from server to client.
+  /// The server repeatedly returns a payload with the size requested by the client.
+  func streamingFromServer(
+    request: ServerRequest.Single<Grpc_Testing_BenchmarkService.Method.StreamingFromServer.Input>
+  ) async throws
+    -> ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingFromServer.Output>
+  {
+    if request.message.responseStatus.isInitialized {
+      try self.checkOkStatus(request.message.responseStatus)
+    }
+    let response = Grpc_Testing_BenchmarkService.Method.StreamingCall.Output.with {
+      $0.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: Int(request.message.responseSize))
+      }
+    }
+    return ServerResponse.Stream { writer in
+      while working.load(ordering: .acquiring) {
+        try await writer.write(response)
+      }
+      return [:]
+    }
+  }
+
+  /// Two-sided unbounded streaming between server to client.
+  /// Both sides send the content of their own choice to the other.
+  func streamingBothWays(
+    request: GRPCCore.ServerRequest.Stream<
+      Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Input
+    >
+  ) async throws
+    -> ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Output>
+  {
+    // The 100 size is used by the other implementations as well.
+    // We are using the same canned response size for all responses
+    // as it is allowed by the spec.
+    let response = Grpc_Testing_BenchmarkService.Method.StreamingCall.Output.with {
+      $0.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: 100)
+      }
+    }
+
+    var serverResponse = ServerResponse.Stream<
+      Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Output
+    >(accepted: .failure(RPCError(code: .internalError, message: "There was a server issue.")))
+
+    try await withThrowingTaskGroup(
+      of: ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Output>?.self
+    ) { group in
+      group.addTask {
+        for try await message in request.messages {
+          if message.responseStatus.isInitialized {
+            try self.checkOkStatus(message.responseStatus)
+          }
+        }
+        _ = self.working.exchange(false, ordering: .acquiring)
+        return nil
+      }
+      group.addTask {
+        return ServerResponse.Stream { writer in
+          while working.load(ordering: .acquiring) {
+            try await writer.write(response)
+          }
+          return [:]
+        }
+      }
+      for try await result in group {
+        guard let result = result else {
+          continue
+        }
+        serverResponse = result
+        group.cancelAll()
+      }
+      group.cancelAll()
+    }
+    return serverResponse
```

Comment:
This is a bit inverted. At the moment the behaviour will be:
- RPC starts
- Task group is created
  - One task runs to consume the request messages
  - The other task returns the response stream immediately
- We then consume the response stream, the first result we'll get will be the server response because it's returned immediately so we cancel the task group (and stop consuming requests)
- We then wait for all child tasks in the task group to finish (this should be fast because one has already finished and the other has been cancelled)
- Then we return the server response so that gRPC can run it.

Instead we should be consuming the request stream within a task group _inside_ the response stream. The rough shape is something like this:

```swift
return ServerResponse.Stream { writer in 
  try await withThrowingTaskGroup(of: Void.self) { group in
    group.addTask { 
      // consume requests 
    } 
    group.addTask { 
      // write responses 
    }
  }
  return [:]
}
```